contracts/FIFOMarket.sol

```diff
@@ -87,65 +100,81 @@ contract FIFOMarket is
     bytes calldata userData,
     bytes calldata
   ) external override {
-    // todo this doesnt seem to revert if the queue is empty
-    // todo take into consideration the fact that the user is sending the amount + 15% fee
+    uint256 certificateAmount = (amount * 100) / (100 + _noriFee);
+    uint256 remainingAmountToFill = certificateAmount;
+
     address recipient = abi.decode(userData, (address)); // todo handle the case where someone invokes this function without operatorData
     require(recipient == address(recipient),"FIFOMarket: Invalid address");
     require(recipient != address(0), "FIFOMarket: Cannot mint to the 0 address");
-    require(msg.sender == address(_nori), "FIFOMarket: This contract can only receive NORI"); // todo verify this can only be invoked by the nori contract
-    uint certificateAmount = (amount * 100) / (100 + _noriFee);
-    uint remainingAmountToFill = certificateAmount;
-    uint256[] memory ids = new uint256[](_queue.length());
-    uint256[] memory amounts = new uint256[](_queue.length());
-    address[] memory suppliers = new address[](_queue.length());
-    for (uint i = 0; i < _queue.length(); i++) {
-      uint removalAmount = _removal.balanceOf(address(this), _queue.at(i));
-      address supplier = _removal.vintage(_queue.at(i)).supplier;
-      if(remainingAmountToFill < removalAmount) {
-        ids[i] = _queue.at(i);
+    // todo verify this can only be invoked by the nori contract
+    require(msg.sender == address(_nori), "FIFOMarket: This contract can only receive NORI");
+    uint256[] memory ids = new uint256[](_queueLength());
+    uint256[] memory amounts = new uint256[](_queueLength());
+    address[] memory suppliers = new address[](_queueLength());
+    for (uint256 i = _queueHeadIndex; i < _queueNextInsertIndex; i++) {
+      uint256 removalAmount = _removal.balanceOf(
+        address(this),
+        _queue[i]
+      );
+      address supplier = _removal.vintage(_queue[i]).supplier;
+      if (remainingAmountToFill < removalAmount) {
+        ids[i] = _queue[i];
         amounts[i] = remainingAmountToFill;
         suppliers[i] = supplier;
         remainingAmountToFill = 0;
-      } else if(remainingAmountToFill >= removalAmount) {
-        if(i == _queue.length() - 1 && remainingAmountToFill > removalAmount){
+      } else if (remainingAmountToFill >= removalAmount) {
```

Comment:
Can we remove the `if` here?  It's true by definition given the preceding `if` clause.  This has the potential to cause bugs / confusion later.